// jsii-docgen/lib/docgen/view/api-reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiReference = void 0;
const classes_1 = require("./classes");
const constructs_1 = require("./constructs");
const enums_1 = require("./enums");
const interfaces_1 = require("./interfaces");
const structs_1 = require("./structs");
/**
 * Render an API reference based on the jsii assembly.
 */
class ApiReference {
    constructor(transpile, assembly, submodule, allSubmodules) {
        let classes;
        let interfaces;
        let enums;
        if (allSubmodules !== null && allSubmodules !== void 0 ? allSubmodules : false) {
            classes = this.sortByName([...assembly.classes, ...flatMap(assembly.allSubmodules, submod => [...submod.classes])]);
            interfaces = this.sortByName([...assembly.interfaces, ...flatMap(assembly.allSubmodules, submod => [...submod.interfaces])]);
            enums = this.sortByName([...assembly.enums, ...flatMap(assembly.allSubmodules, submod => [...submod.enums])]);
        }
        else {
            classes = this.sortByName(submodule ? submodule.classes : assembly.classes);
            interfaces = this.sortByName(submodule ? submodule.interfaces : assembly.interfaces);
            enums = this.sortByName(submodule ? submodule.enums : assembly.enums);
        }
        this.constructs = new constructs_1.Constructs(transpile, classes);
        this.classes = new classes_1.Classes(transpile, classes);
        this.structs = new structs_1.Structs(transpile, interfaces);
        this.interfaces = new interfaces_1.Interfaces(transpile, interfaces);
        this.enums = new enums_1.Enums(transpile, enums);
    }
    /**
     * Generate JSON.
     */
    toJson() {
        return {
            constructs: this.constructs.toJson(),
            classes: this.classes.toJson(),
            structs: this.structs.toJson(),
            interfaces: this.interfaces.toJson(),
            enums: this.enums.toJson(),
        };
    }
    sortByName(arr) {
        return [...arr].sort((s1, s2) => s1.name.localeCompare(s2.name));
    }
}
exports.ApiReference = ApiReference;
function flatMap(xs, fn) {
    return Array.prototype.concat(...xs.map(fn));
}
