// jsii-docgen/lib/docgen/view/class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Class = void 0;
const constants_1 = require("./constants");
const initializer_1 = require("./initializer");
const instance_methods_1 = require("./instance-methods");
const properties_1 = require("./properties");
const static_functions_1 = require("./static-functions");
const schema_1 = require("../schema");
const CONSTRUCT_CLASS = 'constructs.Construct';
class Class {
    static isConstruct(klass) {
        if (klass.fqn === CONSTRUCT_CLASS)
            return true;
        if (!klass.base) {
            return false;
        }
        return this.isConstruct(klass.base);
    }
    constructor(transpile, klass) {
        this.transpile = transpile;
        this.klass = klass;
        if (klass.initializer) {
            this.initializer = new initializer_1.Initializer(transpile, klass.initializer);
        }
        this.instanceMethods = new instance_methods_1.InstanceMethods(transpile, klass.allMethods);
        this.staticFunctions = new static_functions_1.StaticFunctions(transpile, klass.allMethods);
        this.constants = new constants_1.Constants(transpile, klass.allProperties);
        this.properties = new properties_1.Properties(transpile, klass.allProperties);
        this.interfaces = klass.interfaces.map((iface) => this.transpile.type(iface));
        this.transpiled = transpile.class(klass);
    }
    toJson() {
        var _a;
        return {
            initializer: (_a = this.initializer) === null || _a === void 0 ? void 0 : _a.toJson(),
            interfaces: this.interfaces.map((iface) => iface.toJson()),
            instanceMethods: this.instanceMethods.toJson(),
            staticMethods: this.staticFunctions.toJson(),
            constants: this.constants.toJson(),
            properties: this.properties.toJson(),
            fqn: this.transpiled.type.fqn,
            displayName: this.transpiled.type.fqn.split('.').pop(),
            id: this.klass.fqn,
            docs: (0, schema_1.extractDocs)(this.klass.docs),
        };
    }
}
exports.Class = Class;
