// jsii-docgen/lib/docgen/view/interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Interface = void 0;
const instance_methods_1 = require("./instance-methods");
const properties_1 = require("./properties");
const schema_1 = require("../schema");
class Interface {
    static isStruct(iface) {
        return iface.datatype;
    }
    constructor(transpile, iface) {
        this.iface = iface;
        this.transpiled = transpile.interface(iface);
        this.instanceMethods = new instance_methods_1.InstanceMethods(transpile, iface.ownMethods);
        this.properties = new properties_1.Properties(transpile, iface.allProperties);
        this.implementations = iface.allImplementations.map((impl) => transpile.type(impl));
        this.bases = iface.interfaces.map((base) => transpile.type(base));
    }
    toJson() {
        return {
            fqn: this.transpiled.type.fqn,
            displayName: this.transpiled.type.fqn.split('.').pop(),
            id: this.iface.fqn,
            implementations: this.implementations.map((impl) => impl.toJson()),
            interfaces: this.bases.map((base) => base.toJson()),
            instanceMethods: this.instanceMethods.toJson(),
            properties: this.properties.toJson(),
            docs: (0, schema_1.extractDocs)(this.iface.docs),
        };
    }
}
exports.Interface = Interface;
