// jsii-docgen/lib/docgen/view/parameter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parameter = void 0;
const reflect = __importStar(require("jsii-reflect"));
const schema_1 = require("../schema");
class Parameter {
    constructor(transpile, parameter) {
        this.parameter = parameter;
        this.transpiledParam = transpile.parameter(parameter);
        this.transpiledCallable = transpile.callable(parameter.method);
    }
    toJson() {
        var _a;
        const isInitializer = this.parameter.method.kind === reflect.MemberKind.Initializer;
        const methodName = isInitializer ? 'Initializer' : this.transpiledCallable.name;
        const methodId = isInitializer ? 'Initializer' : this.parameter.method.name;
        return {
            fqn: `${this.transpiledParam.parentType.fqn}.${methodName}.parameter.${this.transpiledParam.name}`,
            displayName: this.transpiledParam.name,
            id: `${this.parameter.parentType.fqn}.${methodId}.parameter.${this.parameter.name}`,
            optional: this.transpiledParam.optional === true ? true : undefined, // to save space
            default: (_a = this.parameter.spec.docs) === null || _a === void 0 ? void 0 : _a.default,
            type: this.transpiledParam.typeReference.toJson(),
            docs: (0, schema_1.extractDocs)(this.parameter.docs),
        };
    }
}
exports.Parameter = Parameter;
