// jsii-docgen/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmError = exports.TransliterationError = exports.LanguageNotSupportedError = exports.CorruptedAssemblyError = exports.UnInstallablePackageError = exports.DocGenError = exports.NoSpaceLeftOnDevice = void 0;
// eslint-disable-next-line @typescript-eslint/no-require-imports
const { name } = require('../package.json');
/**
 * The error raised when processing a package fails due to running out of disk
 * space while installing it's dependency closure in a temporary directory. This
 * error cannot be immediately recovered, short of deleting files to make more
 * space availabe, then retrying.
 *
 * Users may perform an `err instanceof NoSpaceLeftOnDevice` test to determine
 * whether this error was raised or not, and cut retry attempts.
 */
class NoSpaceLeftOnDevice extends Error {
    /** @internal */
    constructor(message, stack) {
        super(message);
        this.name = `${name}.${this.constructor.name}`;
        if (this.stack) {
            this.stack = stack;
        }
        else {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.NoSpaceLeftOnDevice = NoSpaceLeftOnDevice;
/**
 * Generic error thrown by the library.
 */
class DocGenError extends Error {
    /** @internal */
    constructor(message, stack) {
        super(message);
        this.name = `${name}.${this.constructor.name}`;
        if (this.stack) {
            this.stack = stack;
        }
        else {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.DocGenError = DocGenError;
/**
 * Raised when docgen is unable to install the given package.
 * This can happen due to invalid dependency clojures for example.
 */
class UnInstallablePackageError extends DocGenError {
}
exports.UnInstallablePackageError = UnInstallablePackageError;
/**
 * Raised when docgen detects corrupted assemblies, preventing it from
 * generating documentation for a specific language.
 * This can happen either due to jsii compiler bugs, or authoring mistakes.
 *
 * For example: https://github.com/aws/jsii/pull/3147
 */
class CorruptedAssemblyError extends DocGenError {
}
exports.CorruptedAssemblyError = CorruptedAssemblyError;
/**
 * Raised when a render is requested for a language the package does not support.
 */
class LanguageNotSupportedError extends DocGenError {
}
exports.LanguageNotSupportedError = LanguageNotSupportedError;
;
/**
 * Raised when snippet transliteration into a target language failed.
 */
class TransliterationError extends DocGenError {
}
exports.TransliterationError = TransliterationError;
;
/**
 * The error raised when `npm` commands fail with an "opaque" exit code,
 * attempting to obtain more information from the commands output.
 */
class NpmError extends DocGenError {
    /** @internal */
    constructor(message, stdout, npmErrorCode) {
        super(message);
        Error.captureStackTrace(this, this.constructor);
        this.stdout = stdout;
        this.npmErrorCode = npmErrorCode;
        this.name = `${name}.${this.constructor.name}${npmErrorCode ? `.${npmErrorCode}` : ''}`;
    }
}
exports.NpmError = NpmError;
