// jsii-docgen/lib/docgen/render/markdown-doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkdownDocument = void 0;
/**
 * Markdown element.
 */
class MarkdownDocument {
    /**
     * Sanitize markdown reserved characters from external input.
     */
    static sanitize(line) {
        let sanitized = line.trim();
        if (line.startsWith('-')) {
            sanitized = sanitized.substring(1, line.length).trim();
        }
        return sanitized;
    }
    /**
     * Remove newlines from markdown.
     */
    static removeNewlines(line) {
        return line.replace(/\n/g, ' ');
    }
    static bold(text) {
        return `**${text}**`;
    }
    static pre(text) {
        // using <code> instead of backticks since this allows links
        return `<code>${text}</code>`;
    }
    static italic(text) {
        return `*${text}*`;
    }
    constructor(options = {}) {
        var _a, _b;
        this.options = options;
        this._lines = new Array();
        this._sections = new Array();
        this.id = (_a = options.id) !== null && _a !== void 0 ? _a : (_b = options.header) === null || _b === void 0 ? void 0 : _b.title;
        this.header = this.formatHeader();
    }
    /**
     * Render a docs element into the markdown.
     */
    docs(docs, language) {
        if (docs.summary) {
            this.lines(MarkdownDocument.sanitize(docs.summary));
            this.lines('');
        }
        if (docs.remarks) {
            this.lines(MarkdownDocument.sanitize(docs.remarks));
            this.lines('');
        }
        if (docs.links) {
            for (const link of docs.links) {
                this.quote(`[${link}](${link})`);
            }
        }
        if (docs.example) {
            if (!language) {
                throw new Error('language must be provided if docs.example has been specified');
            }
            const example = new MarkdownDocument({
                id: `${this.options.id}.example`,
            });
            example.lines(MarkdownDocument.italic('Example'), '');
            example.code(language.toString(), docs.example);
            example.lines('');
            this.section(example);
        }
    }
    table(data) {
        const numColumns = data[0].length;
        const header = data[0];
        const rows = data.slice(1);
        this.lines('| ' + header.map(this.escapePipes).join(' | ') + ' |');
        this.lines('|' + ' --- |'.repeat(numColumns));
        for (const row of rows) {
            this.lines('| ' + row.map(this.escapePipes).join(' | ') + ' |');
        }
        this.lines('');
    }
    quote(line) {
        this.lines(`> ${line}`);
        this.lines('');
    }
    bullet(line) {
        this.lines(`- ${line}`);
    }
    code(language, ...snippet) {
        this.lines(`\`\`\`${language}`, ...snippet, '```');
        this.lines('');
    }
    lines(...lines) {
        this._lines.push(...lines);
    }
    split() {
        this.lines('---');
        this.lines('');
    }
    section(section) {
        this._sections.push(section);
    }
    render(headerSize = 0) {
        var _a, _b;
        const content = [];
        if (this.header) {
            if (headerSize > 6) {
                // headers are mapped to `h1-h6` html elements.
                // passed that, markdown just renders `#` signs.
                // lets see if and when we'll hit this limit.
                throw new Error('Unable to render markdown. Header limit (6) reached.');
            }
            const heading = `${'#'.repeat(headerSize)} ${this.header}`;
            // temporary hack to avoid breaking Construct Hub
            const headerSpan = !!process.env.HEADER_SPAN;
            if (headerSpan) {
                content.push(`${heading} <span data-heading-title="${(_a = this.options.header) === null || _a === void 0 ? void 0 : _a.title}" data-heading-id="${this.id}"></span>`);
            }
            else {
                content.push(`${heading} <a name="${(_b = this.options.header) === null || _b === void 0 ? void 0 : _b.title}" id="${this.id}"></a>`);
            }
            content.push('');
        }
        for (const line of this._lines) {
            content.push(`${line}`);
        }
        for (const section of this._sections) {
            content.push(section.render(headerSize + 1));
        }
        return content.join('\n');
    }
    formatHeader() {
        var _a, _b, _c, _d, _e, _f, _g;
        if (!((_a = this.options.header) === null || _a === void 0 ? void 0 : _a.title)) {
            return undefined;
        }
        let caption = this.options.header.title;
        if ((_c = (_b = this.options.header) === null || _b === void 0 ? void 0 : _b.pre) !== null && _c !== void 0 ? _c : false) {
            caption = `\`${caption}\``;
        }
        if ((_e = (_d = this.options.header) === null || _d === void 0 ? void 0 : _d.strike) !== null && _e !== void 0 ? _e : false) {
            caption = `~~${caption}~~`;
        }
        if ((_f = this.options.header) === null || _f === void 0 ? void 0 : _f.sup) {
            caption = `${caption}<sup>${(_g = this.options.header) === null || _g === void 0 ? void 0 : _g.sup}</sup>`;
        }
        return caption;
    }
    escapePipes(line) {
        return line.replace(/\|/g, '\\|');
    }
}
exports.MarkdownDocument = MarkdownDocument;
/**
 * An empty markdown element.
 */
MarkdownDocument.EMPTY = new MarkdownDocument();
